include/alpaka/kernel/Traits.hpp

```diff
@@ -199,14 +199,41 @@ namespace alpaka
                 static_assert(std::is_same_v<Result, void>, "The TKernelFnObj is required to return void!");
             }
         };
+
+        //! Check that all types are trivially copyable.
+        template<size_t TIdx, typename THead, typename... TTail>
+        struct AssertTriviallyCopyable
+            : AssertTriviallyCopyable<TIdx, THead>
+            , AssertTriviallyCopyable<TIdx + 1, TTail...>
+        {
+        };
+
+        template<size_t TIdx, typename THead>
+        struct AssertTriviallyCopyable<TIdx, THead>
+        {
+            static_assert(std::is_trivially_copyable_v<THead>, "Kernels N-th argument must be trivially copyable!");
+        };
+
+        //! Check that each argument type is trivially copyable
+        //
+        // \tparam TArgs types to be checked
+        template<typename... TArgs>
+        inline void assertArgsTriviallyCopyable()
+        {
+            if constexpr(sizeof...(TArgs))
+            {
+                [[maybe_unused]] auto check = AssertTriviallyCopyable<0, TArgs...>{};
+            }
+        }
```

Comment:
Prefer fold expressions over recursive template expansion. It compiles a lot faster in my experience.

Alternative 1:
```suggestion
template<typename TArg, std::size_t N>
inline void assertArgTriviallyCopyable() {
    static_assert(std::is_trivially_copyable_v<TArg>,
        "Kernels N-th argument must be trivially copyable!");
}

template<typename... TArgs, std::size_t... Is>
inline void assertArgsTriviallyCopyable(std::index_sequence<Is...>)
{
    (assertArgTriviallyCopyable<TArgs, Is>(), ...);
}
```
And then use like:
```c++
assertArgsTriviallyCopyable<std::decay_t<TArgs>...>(std::index_sequence_for<TArgs...>{});
```

Alternative 2:
```suggestion
        template<typename... TArgs>
        inline void assertArgsTriviallyCopyable()
        {
            using namespace boost::mp11;
            mp_for_each<mp_iota_c<sizeof...(TArgs)>>([](auto ic) {
                static_assert(std::is_trivially_copyable_v<mp_at<mp_list<TArgs...>, decltype(ic)>>,
                "Kernels N-th argument must be trivially copyable!");
            });
        }
```

Reply:
I do not want to introduce mp11 in alpaka. I first wrote all with `std:tuple` but run into the problem that this enforce a standard constructor of the types I checked.

Reply:
I choose your first version but this is still producing the issue with the lambda :-(